### PR TITLE
More node + initial mod support and a fix for leaf decay.

### PR DESCRIPTION
**Pull request by leucome**
While working on a texture pack I found luscious. So I thought why not make a pack with variable color. The initial change add few node from Ethereal  and from default.  

The fix for leaf decay use  ~=1 instead of ==0  directly in default function. 
 So the fix is on the mine test game side.  It was merged so now it work as expected. 

There Is a part of the code that I do not know what it's meant for, maybe you can take a look.   Line 194. I added stuff following a similar pattern but it's probably not required. 

Also It's modified so that it can get the heat and humidity from biome data.

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -11,6 +11,8 @@
 
 ]]--
 
+
+
```
Comment (sofar) [inline init.lua]: Drop unrelated whitespace changes, please.

File: init.lua [hunk 2]
```diff
@@ -46,10 +48,15 @@ local function on_construct(pos)
 	local map = minetest.decompress(z)
 
 	local node = minetest.get_node(pos)
-	node.param2 = cmpy(string.byte(map, l + 1), pos.y)
+	
+    if node.param2 ~=1 then     -- Use/preseve 1 as the flag to prevent leaf decay.     
+        node.param2 = cmpy(string.byte(map, l + 1), pos.y)
+    end
```
Comment (sofar) [inline init.lua]: using spaces to indent. Please use tabs.

File: init.lua [hunk 3]
```diff
@@ -59,9 +66,82 @@ minetest.override_item("default:dirt_with_grass", {
 	tiles = {"luscious_grass.png", {name = "default_dirt.png", color = "white"}, "luscious_grass.png"},
 	overlay_tiles = {"", "", {name = "luscious_dirt_overlay.png", color = "white"}},
 	place_param2 = 136,
+	drop  = "default:dirt",
+	on_construct = on_construct,
+})
+
```
Comment (sofar) [inline init.lua]: 1 line empty space here instead.

File: init.lua [hunk 4]
```diff
@@ -46,10 +48,15 @@ local function on_construct(pos)
...
-	node.param2 = cmpy(string.byte(map, l + 1), pos.y)
+	
+    if node.param2 ~=1 then     -- Use/preseve 1 as the flag to prevent leaf decay.     
+        node.param2 = cmpy(string.byte(map, l + 1), pos.y)
+    end
 	minetest.swap_node(pos, node)
 end
 
+
+
```
Comment (sofar) [inline init.lua]: same here, keep the changes just to the changed code please.

File: init.lua [hunk 5]
```diff
@@ -59,9 +66,82 @@ minetest.override_item("default:dirt_with_grass", {
...
+minetest.override_item("default:sand", {
+	paramtype2 = "color",
+	drawtype = "color",
+	palette_index = 136,
+	color = "#f3dec3ff",
+	palette = "luscious_default_sand_palette.png",
+	tiles = {"luscious_default_sand.png", {name = "default_sand.png", color = "white"}, "luscious_default_sand.png"},
+--	overlay_tiles = {"", "", {name = "luscious_dirt_overlay.png", color = "white"}},
```
Comment (sofar) [inline init.lua]: because there is no overlay tile, remove this line

File: init.lua [hunk 6]
```diff
@@ -59,9 +66,82 @@ minetest.override_item("default:dirt_with_grass", {
...
+	place_param2 = 136,
+	drop  = "ethereal:dry_dirt",
 	on_construct = on_construct,
 })
 
+
+
+end
+
+-----------------------------------------------------------------------------------------------------------
```
Comment (sofar) [inline init.lua]: No need for `-------------` lines either.

File: init.lua [hunk 7]
```diff
@@ -80,6 +160,7 @@ for _, v in pairs({
 		color = "#8afb94ff",
 		tiles = {tile},
 		place_param2 = 136,
+		groups = {snappy = 3, flammable = 2, leaves = 1, leafdecay = 1},
```
Comment (sofar) [inline init.lua]: this shouldn't be needed. We don't want to override unrelated properties of the nodes.

File: init.lua [hunk 8]
```diff
@@ -108,12 +190,18 @@ for _, v in pairs({
 		on_construct = on_construct,
 	})
 end
+
+--- I have absolutly no idea what is the purpose but I aplied it to other stuff and it work.
 for i = 1, 5 do
-	minetest.register_alias_force("default:dry_grass_" .. i, "default:grass_" .. i)
+	minetest.register_alias_force("default:dry_grass_" .. i, "default:grass_" .. i , "default:desert_" .. i , "ethereal:dry_" .. i, "ethereal:prairie_" .. i )
```
Comment (sofar) [inline init.lua]: This is definitely wrong. register_alias_force makes all dry_grass into grass so that they can be colored using the same algorithm. This is needed for when Biome A has a "dry" variant for Biome B's nodes.

There can be only 2 parameters.

So if you want to do this for ethereal dry grass, then you need 1 extra line
if you want to do this for the 2 sand types, you need another extra call

It doesn't break anything because the API throws out parameter 3, 4 etc.

File: init.lua [hunk 9]
```diff
@@ -108,12 +190,18 @@ for _, v in pairs({
...
 end
 
 -- content ids
 local cn = {
+	["ethereal:dry_dirt"] = 1,
+	["ethereal:prairie_dirt"] = 1,
+    ["default:sand"] = 1,
```
Comment (sofar) [inline init.lua]: Tabs/spaces again

File: init.lua [hunk 10]
```diff
@@ -12,46 +12,26 @@
...
-	if y > 0 then
-		local h1 = p2 % 16
-		local h2 = math.floor(p2 / 16) * 16
-
-		h1 = h1 - math.max(math.min(h1, math.floor(y / 16)), 0)
-		return math.max(1, h1 + h2)
-	else
-		return math.max(1, p2)
-	end
+local function chose_color(pos)
```
Comment (sofar) [inline init.lua]: tabs and spaces again.

When you modify an existing project, you should adopt the code style, not change it.

File: init.lua [hunk 11]
```diff
@@ -12,46 +12,26 @@
...
+    -- Calculate color from biome heat and humidity.
+    local biome_data = minetest.get_biome_data(pos)
+    local hum1 = biome_data.humidity
+    local temp1 = biome_data.heat
+    local vh = math.floor(math.min(math.max(math.floor(hum1), 0), 100) / 6.6)
+	local vt = math.floor(math.min(math.max(math.floor(temp1), 0), 100) / 6.6)
+	local c1 = (vt + (vh * 16))
+    --minetest.chat_send_all(temp1)
```
Comment (sofar) [inline init.lua]: remove debug printouts.

File: init.lua [hunk 12]
```diff
@@ -12,46 +12,26 @@
...
-		return math.max(1, p2)
-	end
+local function chose_color(pos)
+    -- Calculate color from biome heat and humidity.
+    local biome_data = minetest.get_biome_data(pos)
+    local hum1 = biome_data.humidity
+    local temp1 = biome_data.heat
+    local vh = math.floor(math.min(math.max(math.floor(hum1), 0), 100) / 6.6)
+	local vt = math.floor(math.min(math.max(math.floor(temp1), 0), 100) / 6.6)
+	local c1 = (vt + (vh * 16))
```
Comment (sofar) [inline init.lua]: extra parenthesis pair can be removed.

I'd Prefer:
```
local c1 = vh * 16 + vt
```
since no parentheses needed

File: init.lua [hunk 13]
```diff
@@ -12,46 +12,26 @@
...
+    --minetest.chat_send_all(temp1)
+    return c1
```
Comment (sofar) [inline init.lua]: why not `return (vh * 16 + vt)` ?

File: init.lua [hunk 14]
```diff
@@ -12,46 +12,26 @@
...
 
-	local z = f:read("*a")
-	f:close()
-	local map = minetest.decompress(z)
 
-	local node = minetest.get_node(pos)
-	
-    if node.param2 ~=1 then     -- Use/preseve 1 as the flag to prevent leaf decay.     
-        node.param2 = cmpy(string.byte(map, l + 1), pos.y)
-    end
```
Comment (sofar) [inline init.lua]: You're throwing out all the caching here. It would be likely good for performance if you would still make the cache and use it instead of just calling the API function, since caching is cheap and very efficient here (since every vertical Y column has only a 16x16 cache block).

File: init.lua [hunk 15]
```diff
@@ -12,46 +12,26 @@
...
+local function on_construct(pos)
+	-- get chunk from pos and apply color
+    local node = minetest.get_node(pos)
+    node.param2 = chose_color(pos)
```
Comment (sofar) [inline init.lua]: tabs/spaces again

File: init.lua [hunk 16]
```diff
@@ -79,7 +59,7 @@ minetest.override_item("default:desert_sand", {
 	palette_index = 136,
 	color = "#c0825dff",
 	palette = "luscious_desert_sand_palette.png",
-	tiles = {"luscious_desert_sand.png", {name = "luscious_desert_sand", color = "white"}, "luscious_desert_sand.png"},
+	tiles = {"luscious_desert_sand.png", {"luscious_desert_sand.png"}, "luscious_desert_sand.png"},
```
Comment (sofar) [inline init.lua]: if you're doing this then the `{}` are also not needed here.

File: init.lua [hunk 17]
```diff
@@ -177,7 +167,13 @@ for _, v in pairs({
 	"default:grass_4",
 	"default:grass_5",
 	"default:junglegrass",
-}) do
+}) do            local p = vector.new(x, y, z)
```
Comment (sofar) [inline init.lua]: keep the line break in between (code style adopt, not change)

File: init.lua [hunk 18]
```diff
@@ -177,7 +167,13 @@ for _, v in pairs({
...
+            local biome_data = minetest.get_biome_data(p)
+            local hum1 = biome_data.humidity
+            local temp1 = biome_data.heat
+            local vh = math.floor(math.min(math.max(math.floor(hum1), 0), 100) / 6.6)
```
Comment (sofar) [inline init.lua]: tabs/spaces here

File: init.lua [hunk 19]
```diff
@@ -177,7 +167,13 @@ for _, v in pairs({
...
+            local vt = math.floor(math.min(math.max(math.floor(temp1), 0), 100) / 6.6)
+            local v1 = (vt + (vh * 16))
```
Comment (sofar) [inline init.lua]: duplicate code, make a function

File: init.lua [hunk 20]
```diff
@@ -187,17 +183,21 @@ for _, v in pairs({
 		color = "#7dea1eff",
 		tiles = {tile},
 		place_param2 = 136,
+        visual_scale = 2.3,
```
Comment (sofar) [inline init.lua]: where did this come from?

File: init.lua [hunk 21]
```diff
@@ -222,45 +222,34 @@ for k, _ in pairs(cn) do
...
-		local h2 = humiditymap[i]
-		h1 = math.floor(math.min(math.max(math.floor(h1), 0), 100) / 6.6)
-		h2 = math.floor(math.min(math.max(math.floor(h2), 0), 100) / 6.6)
-		map = map .. string.char(h1 + (h2 * 16))
-	end
-	local p = string.format("%d", minetest.hash_node_position(v))
 
-	local f = assert(io.open(wp .. "/" .. p, "w"), wp .. "/" .. p)
-	f:write(minetest.compress(map))
-	f:close()
```
Comment (sofar) [inline init.lua]: same cache comment: you're throwing away a good optimization here.

File: init.lua [hunk 22]
```diff
@@ -222,45 +222,34 @@ for k, _ in pairs(cn) do
...
 	for z = minp.z, maxp.z do
 	for y = minp.y, maxp.y do
 		local vi = area:index(minp.x, y, z)
 		for x = minp.x, maxp.x do
 			local vv = (x - minp.x) + ((z - minp.z) * chunksize)
 			if cs[data[vi]] then
-				local mv = cmpy(string.byte(map, vv + 1), y)
-				p2data[vi] = mv
+            local p = vector.new(x, y, z)
+            local mv = chose_color(p)
```
Comment (sofar) [inline init.lua]: `local mv = chose_color(vector.net(x, y, z))`

File: init.lua [hunk 23]
```diff
@@ -222,45 +222,34 @@ for k, _ in pairs(cn) do
...
+            p2data[vi] = mv                
 			end
 			vi = vi + 1
 		end
 	end
 	end
-	vm:set_param2_data(p2data)
-	vm:write_to_map()
+vm:set_param2_data(p2data)
```
Comment (sofar) [inline init.lua]: indentation is all over the place. here there is none.

File: init.lua [hunk 24]
```diff
@@ -270,7 +259,7 @@ local function update_p2(pos, size)
 		local p = vector.new(x, y, z)
 		local node = minetest.get_node(p)
 		if cn[node.name] then
-			on_construct(p)
+        on_construct(p)
```
Comment (sofar) [inline init.lua]: again unrelated whitespace changes - don't do this

**Comment by sofar** [comment]
can you rebase/squash this into the following individual changes:
1. using get_biome APi change (only)
2. fix leafdecay
3. add new nodes (sand)
4. add ethereal

You should end up with 3-4 commits, not 8. Please look at `git rebase -i f1f2cd4d4e45c62040740f80760cd4ca1d550edf` for instance.

File: init.lua (hunk 10 above)
Comment (leucome) [reply init.lua]: Turn out that  my software was converting every new tab to space.

File: init.lua (hunk 20 above)
Comment (leucome) [reply init.lua]: It allow grass to overlap each other a little when seen at ground level. There is wider grass textures  to preview the concept.  Currently 64px early sketch. But If you like I'll provide final one likely to be 32px to match the visual style of the default game.

File: init.lua (hunk 14 above)
Comment (leucome) [reply init.lua]: Yes , I kinda overlooked the performance side I thought it was principllay meant to store the information for later use.  
I wonder how much slower it is.  Is there a built in benchmark for generation to compare the two?

File: init.lua (hunk 7 above)
Comment (leucome) [reply init.lua]: Of course preserving the normal behaviors of leafdecay is really important.   Said with a little bit of sarcasm.

**Comment by sofar** [comment]
Why did you delete everything?

**Comment by leucome** [comment]
To redo the commit by individual change/feature. I used the web interface so most of the commit were per files instead of specific feature.
